depthcharge/transformers/peptides.py

```diff
@@ -0,0 +1,289 @@
+        The number of tokens used to tokenize peptide sequences.
+    d_model : int
+        The latent dimensionality to represent the amino acids in a peptide
+        sequence.
+    nhead : int, optional
+        The number of attention heads in each layer. ``d_model`` must be
+        divisible by ``nhead``.
+    dim_feedforward : int, optional
+        The dimensionality of the fully connected layers in the Transformer
+        layers of the model.
+    n_layers : int, optional
+        The number of Transformer layers.
+    dropout : float, optional
+        The dropout probability for all layers.
+    positional_encoder : PositionalEncoder or bool, optional
+        The positional encodings to use for the amino acid sequence. If
+        ``True``, the default positional encoder is used. ``False`` disables
+        positional encodings, typically only for ablation tests.
+    max_charge : int, optional
+        The maximum charge state for peptide sequences.
+    """
+
+    def __init__(
+        self,
+        n_tokens: int | PeptideTokenizer,
+        d_model: int = 128,
+        nhead: int = 8,
+        dim_feedforward: int = 1024,
+        n_layers: int = 1,
+        dropout: float = 0,
+        positional_encoder: PositionalEncoder | bool = True,
+        max_charge: int = 5,
+    ) -> None:
+        """Initialize a PeptideEncoder."""
+        super().__init__(
+            n_tokens=n_tokens,
+            d_model=d_model,
+            positional_encoder=positional_encoder,
+            max_charge=max_charge,
+        )
+
+        # The Transformer layers:
+        layer = torch.nn.TransformerEncoderLayer(
+            d_model=d_model,
+            nhead=nhead,
+            dim_feedforward=dim_feedforward,
+            batch_first=True,
+            dropout=dropout,
+        )
+
+        self.transformer_encoder = torch.nn.TransformerEncoder(
+            layer,
+            num_layers=n_layers,
+        )
+
+    def forward(
+        self,
+        tokens: torch.Tensor,
+        charges: torch.Tensor,
+    ) -> tuple[torch.Tensor, torch.Tensor]:
+        """Predict the next amino acid for a collection of sequences.
+
+        Parameters
+        ----------
+        tokens : torch.Tensor of size (batch_size, peptide_length)
+            The integer tokens describing each peptide sequence, padded
+            to the maximum peptide length in the batch with 0s.
+        charges : torch.Tensor of size (batch_size,)
+            The charge state of each peptide.
+
+        Returns
+        -------
+        latent : torch.Tensor of shape (n_sequences, len_sequence, d_model)
+            The latent representations for the spectrum and each of its
+            peaks.
+        mem_mask : torch.Tensor
+            The memory mask specifying which elements were padding in X.
+        """
+        # Encode everything:
+        encoded = self.aa_encoder(tokens)
+        charges = self.charge_encoder(charges - 1)[:, None]
```

Comment:
**question:** Why is there a minus 1 here?

Reply:
The embedding layer had a number of entries equal to the max charge state. However, I changed this just now to be max charge + 1, so that one could also model neutrals effectively.